File: polars/expr_temporal.go
package polars

import (
	pb "github.com/isesword/polars-go-bridge/proto"
)

// DtYear extracts the year from a Date/Datetime expression.
func (e Expr) DtYear() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtYear{
				DtYear: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtMonth extracts the month from a Date/Datetime expression.
func (e Expr) DtMonth() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtMonth{
				DtMonth: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtDay extracts the day from a Date/Datetime expression.
func (e Expr) DtDay() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtDay{
				DtDay: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtWeekday extracts the ISO weekday from a Date/Datetime expression.
func (e Expr) DtWeekday() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtWeekday{
				DtWeekday: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtHour extracts the hour from a Datetime/Time expression.
func (e Expr) DtHour() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtHour{
				DtHour: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtMinute extracts the minute from a Datetime/Time expression.
func (e Expr) DtMinute() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtMinute{
				DtMinute: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtSecond extracts the second from a Datetime/Time expression.
func (e Expr) DtSecond() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtSecond{
				DtSecond: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtMonthStart rolls backward to the first day of the month.
func (e Expr) DtMonthStart() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtMonthStart{
				DtMonthStart: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// DtMonthEnd rolls forward to the last day of the month.
func (e Expr) DtMonthEnd() Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_DtMonthEnd{
				DtMonthEnd: &pb.TemporalUnary{Expr: e.inner},
			},
		},
	}
}

// StrToDate parses a string column into a Date using the provided format.
func (e Expr) StrToDate(format string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrToDate{
				StrToDate: &pb.TemporalParse{
					Expr:   e.inner,
					Format: format,
				},
			},
		},
	}
}

// StrToDatetime parses a string column into a Datetime using the provided format.
func (e Expr) StrToDatetime(format string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrToDatetime{
				StrToDatetime: &pb.TemporalParse{
					Expr:   e.inner,
					Format: format,
				},
			},
		},
	}
}

// StrToTime parses a string column into a Time using the provided format.
func (e Expr) StrToTime(format string) Expr {
	return Expr{
		inner: &pb.Expr{
			Kind: &pb.Expr_StrToTime{
				StrToTime: &pb.TemporalParse{
					Expr:   e.inner,
					Format: format,
				},
			},
		},
	}
}

File: rust/src/expr_temporal.rs
use polars::prelude::*;

use crate::error::BridgeError;
use crate::executor::build_expr;
use crate::proto;

pub fn build_temporal_expr(kind: &proto::expr::Kind) -> Option<Result<Expr, BridgeError>> {
    use proto::expr::Kind;

    match kind {
        Kind::DtYear(func) => Some(build_unary(func, "DtYear", |expr| expr.dt().year())),
        Kind::DtMonth(func) => Some(build_unary(func, "DtMonth", |expr| expr.dt().month())),
        Kind::DtDay(func) => Some(build_unary(func, "DtDay", |expr| expr.dt().day())),
        Kind::DtWeekday(func) => Some(build_unary(func, "DtWeekday", |expr| expr.dt().weekday())),
        Kind::DtHour(func) => Some(build_unary(func, "DtHour", |expr| expr.dt().hour())),
        Kind::DtMinute(func) => Some(build_unary(func, "DtMinute", |expr| expr.dt().minute())),
        Kind::DtSecond(func) => Some(build_unary(func, "DtSecond", |expr| expr.dt().second())),
        Kind::DtMonthStart(func) => Some(build_unary(func, "DtMonthStart", |expr| {
            expr.dt().month_start()
        })),
        Kind::DtMonthEnd(func) => Some(build_unary(func, "DtMonthEnd", |expr| {
            expr.dt().month_end()
        })),
        Kind::StrToDate(parse) => Some(build_str_to_date(parse)),
        Kind::StrToDatetime(parse) => Some(build_str_to_datetime(parse)),
        Kind::StrToTime(parse) => Some(build_str_to_time(parse)),
        _ => None,
    }
}

fn build_unary<F>(func: &proto::TemporalUnary, name: &str, op: F) -> Result<Expr, BridgeError>
where
    F: FnOnce(Expr) -> Expr,
{
    let expr = build_inner_expr(&func.expr, name)?;
    Ok(op(expr))
}

fn build_str_to_date(parse: &proto::TemporalParse) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&parse.expr, "StrToDate")?;
    let options = StrptimeOptions {
        format: Some(parse.format.clone().into()),
        strict: true,
        exact: true,
        cache: true,
    };
    Ok(expr.str().to_date(options))
}

fn build_str_to_datetime(parse: &proto::TemporalParse) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&parse.expr, "StrToDatetime")?;
    let options = StrptimeOptions {
        format: Some(parse.format.clone().into()),
        strict: true,
        exact: true,
        cache: true,
    };
    Ok(expr.str().to_datetime(None, None, options, lit("raise")))
}

fn build_str_to_time(parse: &proto::TemporalParse) -> Result<Expr, BridgeError> {
    let expr = build_inner_expr(&parse.expr, "StrToTime")?;
    let options = StrptimeOptions {
        format: Some(parse.format.clone().into()),
        strict: true,
        exact: true,
        cache: true,
    };
    Ok(expr.str().to_time(options))
}

fn build_inner_expr(expr: &Option<Box<proto::Expr>>, name: &str) -> Result<Expr, BridgeError> {
    let expr = expr
        .as_ref()
        .ok_or_else(|| BridgeError::PlanSemantic(format!("{name} has no expr")))?;
    build_expr(expr.as_ref())
}
